contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,154 @@
+    // the network ID of an L3. These chain IDs are used as the key in this mapping because network IDs are enforced to be unique. Since we require the chain
+    // ID to be sent along with a message or token relay, ForwarderInterface's relay functions include an extra field, `destinationChainId`, when compared to the
+    // relay functions of `AdapterInterface`.
+    mapping(uint256 => address) chainAdapters;
+
+    event ChainAdaptersUpdated(uint256 indexed destinationChainId, address l2Adapter);
+    event SetXDomainAdmin(address indexed crossDomainAdmin);
+    event DestinationChainTokensUpdated(
+        address indexed target,
+        address indexed baseToken,
+        address indexed destinationChainToken
+    );
+
+    error InvalidCrossDomainAdmin();
+    error RelayMessageFailed();
+    error RelayTokensFailed(address baseToken);
+    // Error which is triggered when there is no adapter set in the `chainAdapters` mapping.
+    error UninitializedRoute();
+
+    /*
+     * @dev Cross domain admin permissioning is implemented specifically for each L2 that this contract is deployed on, so this base contract
+     * simply prescribes this modifier to protect external functions using that L2's specific admin permissioning logic.
+     */
+    modifier onlyAdmin() {
+        _requireAdminSender();
+        _;
+    }
+
+    /**
+     * @notice Constructs the Forwarder contract.
+     * @dev _disableInitializers() restricts anybody from initializing the implementation contract, which if not done,
+     * may disrupt the proxy if another EOA were to initialize it.
+     */
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Initializes the forwarder contract.
+     * @param _crossDomainAdmin L1 address of the contract which can send root bundles/messages to this forwarder contract.
+     */
+    function __Forwarder_init(address _crossDomainAdmin) public onlyInitializing {
+        __UUPSUpgradeable_init();
+        _setCrossDomainAdmin(_crossDomainAdmin);
+    }
+
+    /**
+     * @notice Sets a new cross domain admin for this contract.
+     * @param _newCrossDomainAdmin L1 address of the new cross domain admin.
+     */
+    function setCrossDomainAdmin(address _newCrossDomainAdmin) external onlyAdmin {
+        _setCrossDomainAdmin(_newCrossDomainAdmin);
+    }
+
+    /**
+     * @notice Maps a new destination chain ID to an adapter contract which facilitates bridging to that chain.
+     * @param _destinationChainId The chain ID of the target network.
+     * @param _l2Adapter Contract address of the adapter which interfaces with the L2-L3 bridge.
+     * @dev Actual bridging logic is delegated to the adapter contract so that the forwarder can function irrespective of the "flavor" of
+     * L3 (e.g. ArbitrumOrbit, OpStack, etc.).
+     */
+    function updateRoute(uint256 _destinationChainId, address _l2Adapter) external onlyAdmin {
```

Comment:
nit should this be renamed to updateAdapter?

Reply:
Yes, it should have been. Fixed here https://github.com/across-protocol/contracts/pull/609/commits/6298cc69fd08459d75735bb53cb6f1af66f95e1b